content: draft: source track: cleanup objective section (#1373)
slsa-framework/slsa: docs/spec/draft/source-requirements.md

# SLSA Source Track

> ## Outstanding TODOs

Author: removed, none of the other sections are using this technique now.

> The primary purpose of the SLSA Source track is to provide evidence that a repository revision was created following the expected process.

Reviewer: I think we should probably try to highlight the trustworthiness first (that's the objective), meanwhile the (a?) mechanism we use is by ensuring the revisions followed the expected process.

Author: That makes sense -- "trustworthy evidence" maybe?

Reviewer: What about

```suggestion
The primary purpose of the SLSA Source track is to provide consumers with increasing levels of trust in the source code they consume.
```

We can get into revisions, processes, and evidence later.

It describes increasing levels of trustworthiness and completeness of a revision's <dfn>provenance</dfn>. 

> In this track, provenance describes how a source revision came to exist.

Author: I like the phrasing used on the build-env track: https://github.com/slsa-framework/slsa/blob/f671d6ce4f286557dd79c93ed4b5ecd07e1f6641/docs/spec/draft/attested-build-env-levels.md?plain=1#L28C19-L29C24

Reviewer: Hmm, considering we only require the document 'provenance' at level 3 I'm somewhat concerned that we're overloading terms.

_Here_ we're talking about provenance generally, _there_ we're talking about a specific document.

We can probably address this by:

1. Being very careful when we talk about the provenance document(s) to always talk about "provenance _attestations_"
2. We can use a different work for the document entirely
3. We can drop the general purpose use of the word 'provenance' (I don't think it's used that broadly here?)
4. Other?

We can also probably track this as a separate issue and not block this PR on its resolution.

Reviewer: @zachariahcox do you think this thread has been resolved in some way?  It still looks the same to me.

> It describes increasing levels of trustworthiness and completeness of a revision's <dfn>provenance</dfn>.
> In this track, provenance describes how a source revision came to exist.

Reviewer: ```suggestion
It describes increasing levels of trustworthiness and completeness of how a source revision was produced.
```

> The Source track is scoped to single revisions of repositories managed by single organizations. 

Author: I tried to simplify it, but it might need more. 

I think the key ideas here are: 
1. provenance is issued on a per-revision basis (not per branch or per repo)
2. "correctness" / expectations are set by the owning organization
3. (optional I guess) there must be exactly one organization in charge of a repo? I'm not sure if we need this requirement anymore. I think we just need "it's a revision of a repo where some org is responsible for it"

> The Source track is scoped to single revisions of repositories managed by single organizations. 

Author: Sigh. That's a good point Mr. Copilot.

Actually, maybe we should avoid the "is scoped to" wording. 

I think we really mean "the spec allows for writing provenance about _revisions_, not whole repos."

> Source provenance applies to a single revision of a repository, not the entire repository.

Author: ```suggestion
Source provenance attestations apply to a single revision of a repository, not the entire repository.
```

> The Source track is scoped to single revisions of repositories managed by single organizations.
> The organization determines the intent of the software in the repository, the expected process for the creation of new revisions, and administers technical controls to enforce the process.

Author: ```suggestion
Source provenance applies to a single revision of a repository, not the entire repository.
The expected process for creating a new revision is determined solely by that repository's owner (the organization) who also determines the intent of the software in the repository and administers technical controls to enforce the process.
```

> The organization determines the intent of the software in the repository, the exected process for the creation of new revisions, and administers technical controls to enforce the process.

Author: edited this a bit. 

According to the lead in
1. provenance is a bunch of facts. 
1. Orgs determine what needs to happen, and they pick certain source track techniques to improve the trustworthiness of their process claims.
1. If they have certain sets of controls and techniques, they are eligible for SLSA source summary VSAs.

> Consumers can review source provenance attestations to verify whether a particular revision meets their standards.

Author: this might not be enough, and we might still need the summary concept here.

Reviewer: "but consumers who issue VSAs do consume source provenance."

I've been considering the entity that issues VSAs as being part of the SCS and not a separate consumer (which might actually explain some of the disagreements we've had :p ).  ([reference](https://slsa.dev/spec/draft/source-requirements#definitions:~:text=%E2%9C%93-,Source%20Summary%20Attestations,-The%20SCS%20MUST))

I bet we need a diagram or something that calls out the various parts of an SCS to make this more clear.

Author: Yes. I think it may be normal that parts the SCS are owned by different teams / orgs / producers. This implies that all together the combo of tools represent "the scs" used by "the producer" for "a repo" but those tools may have internal dependencies on each other that matter.

Reviewer 2: > this might not be enough. end-users should probably use the summary attestations, but consumers who issue VSAs do consume source provenance.

I think the sentence in the PR now is good. While most consumers _should_ use VSAs, the ability to independently use source provenance at higher levels should be available to all consumers, not just those producing VSAs.

> I bet we need a diagram or something that calls out the various parts of an SCS to make this more clear.

big +1! filed https://github.com/slsa-framework/slsa/issues/1377 to track this.

Reviewer: > While most consumers should use VSAs, the ability to independently use source provenance at higher levels should be available to all consumers, not just those producing VSAs.

Yup, didn't mean to exclude that and the spec does require provenance be made available to them.  Do you think that needs to be called out in the objective? 

Maybe in the objective we just say something like

```suggestion
Consumers can review attestations to verify whether a particular revision meets their standards.
```

| Source Provenance | Information about how a revision came to exist, where it was hosted, when it was generated, what process was used, who the contributors were, and which parent revisions preceded it.
| Repository / Repo | A uniquely identifiable instance of a VCS. The repository controls access to the Source in the VCS. The objective of a repository is to reflect the intent of the organization that controls it.
| Branch | A named, moveable, pointer to a revision that tracks development in the named context over time. Branches may be modified to point to different revisions by authorized actors. Different branches may have different security requirements.
| Tag | A named pointer to a revision that does not typically move. Similar to branches, tags may be modified by authorized actors. Tags are often used by producers to indicate a more permanent name for a revision.
| Change | A set of modifications to the source in a specific context. A change can be proposed and reviewed before being accepted.
| Change History | A record of the history of revisions that preceded a specific revision.

> | Push / upload / publish | When an actor adds or modifies the Source, Branches or Tags in the repository.

Author: avoid use of new term "content"

> | Push / upload / publish | When an actor adds a new revision to the repository. Typically also modifies a branch to point to the new revision.

Author: removed the reiteration of authentication. 
the verbs here should be "add" and "review" not "authenticates" I think.

> | Review / approve / vote | When an actor uses a change review tool to comment upon, endorse, or reject a source change proposal.

Reviewer: yes I think that's a good change, when I see "review" I think specifically about the thing that people use to view the diffs and approve.

> | Review / approve / vote | When an actor uses a change management tool to comment upon, endorse, or reject a source change proposal.

Author: @TomHennen I went ahead and made this change 👍

Reviews SHOULD cover, at least, security relevant properties of the code.

**[Final revision approved]** This requirement applies to the final revision
submitted. I.e. if additional changes are made during the review process, those changes MUST
be reviewed as well.

**[Context-specific approvals]** Approvals are for a specific context, such as a
repo + branch in git. Moving fully reviewed content from one context to another
still requires review. The exact definition of “context” depends on the project,

> and this does not preclude well-understood automatic merges, such as cutting a 

Author: removing the word review-less -- there may be many policies that are bypassed automatically in this way.

release branch.

**[Trusted Robot Contributions]** An organization MAY choose to grant a Trusted
Robot a perpetual exception to a policy (e.g. a bot may be able to merge a change

> that has not been reviewed by two parties). 

Author: solidly into scope creep here, lol.
In the previous wording, it wasn't clear if a bot could bypass other policies too. 
We should probably add a rider about "please document any actors with perpetual exceptions to stuff"

Reviewer 2: I think the rider makes a lot of sense, I would even say it should be a MUST for the organization.

Do you want to add it to this PR or would it make sense to update this independently? I lean towards the latter since it doesn't seem like a copy edit anymore.

Reviewer: Once we start talking about org defined controls it's going to become very difficult to document all that stuff.

We're currently not even requiring them to document how the controls are implemented in the first place: https://slsa.dev/spec/draft/source-requirements#:~:text=Specify%20control%20expectations

With each new 'MUST' requirement I ask myself "how would an implementation prove that this was done".  The more requirements that harder implementations become.

I'd suggest we leave things as is and perhaps as something to pursue in a later revision.
